nomad/host_volume_endpoint.go

```diff
@@ -0,0 +1,527 @@
+		return err
+	}
+	v.srv.MeasureRPCRate("host_volume", structs.RateMetricRead, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "host_volume", "get"}, time.Now())
+
+	allowVolume := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeRead)
+	aclObj, err := v.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+	if !allowVolume(aclObj, args.RequestNamespace()) {
+		return structs.ErrPermissionDenied
+	}
+
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, store *state.StateStore) error {
+
+			vol, err := store.HostVolumeByID(ws, args.Namespace, args.ID, true)
+			if err != nil {
+				return err
+			}
+
+			reply.Volume = vol
+			if vol != nil {
+				reply.Index = vol.ModifyIndex
+			} else {
+				index, err := store.Index(state.TableHostVolumes)
+				if err != nil {
+					return err
+				}
+
+				// Ensure we never set the index to zero, otherwise a blocking
+				// query cannot be used.  We floor the index at one, since
+				// realistically the first write must have a higher index.
+				if index == 0 {
+					index = 1
+				}
+				reply.Index = index
+			}
+			return nil
+		}}
+	return v.srv.blockingRPC(&opts)
+}
+
+func (v *HostVolume) List(args *structs.HostVolumeListRequest, reply *structs.HostVolumeListResponse) error {
+	authErr := v.srv.Authenticate(v.ctx, args)
+	if done, err := v.srv.forward("HostVolume.List", args, args, reply); done {
+		return err
+	}
+	v.srv.MeasureRPCRate("host_volume", structs.RateMetricList, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "host_volume", "list"}, time.Now())
+
+	aclObj, err := v.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+
+	ns := args.RequestNamespace()
+
+	sort := state.SortOption(args.Reverse)
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, store *state.StateStore) error {
+
+			var iter memdb.ResultIterator
+			var err error
+
+			switch {
+			case args.NodeID != "":
+				iter, err = store.HostVolumesByNodeID(ws, args.NodeID, sort)
+			case args.NodePool != "":
+				iter, err = store.HostVolumesByNodePool(ws, args.NodePool, sort)
+			default:
+				iter, err = store.HostVolumes(ws, sort)
+			}
+			if err != nil {
+				return err
+			}
+
+			// Generate the tokenizer to use for pagination using namespace and
+			// ID to ensure complete uniqueness.
+			tokenizer := paginator.NewStructsTokenizer(iter,
+				paginator.StructsTokenizerOptions{
+					WithNamespace: true,
+					WithID:        true,
+				},
+			)
+
+			filters := []paginator.Filter{
+				paginator.GenericFilter{
+					Allow: func(raw any) (bool, error) {
+						vol := raw.(*structs.HostVolume)
+						if !strings.HasPrefix(vol.Name, args.Prefix) &&
+							!strings.HasPrefix(vol.ID, args.Prefix) {
+							return false, nil
```

Comment:
just noting that this implicitly checks `if args.Prefix != ""`, i.e. an empty `Prefix` does not filter things out.

Reply:
Right. Maybe that's surprising though and deserves a comment